Example/metro.config.js

```diff
@@ -35,42 +42,56 @@ const config = {
       return acc;
     }, {}),
 
-    // Since we use react-naviation as submodule it comes with it's own node_modules. While loading
+    nodeModulesPaths: [path.join(__dirname, '../../')],
+
+    // Since we use react-navigation as submodule it comes with it's own node_modules. While loading
     // react-navigation code, due to how module resolution algorithms works it seems that its node_modules
     // are consulted first, resulting in double-loaded packages (so doubled react, react-native and other package instances) leading
     // to various errors. To mitigate this we define below custom request resolver, hijacking requests to conflicting modules and manually
     // resolving appropriate files. **Most likely** this can be achieved by proper usage of blockList but I found this method working ¯\_(ツ)_/¯
     resolveRequest: (context, moduleName, platform) => {
-      if (moduleName === 'react' || moduleName === 'react-native') {
+      if (moduleName.startsWith('@react-navigation')) {
+        // For some reason, react-navigation packages don't want to resolve from
+        // the project's node_modules, so we need to use standard Metro resolver.
+        return context.resolveRequest(context, moduleName, platform);
+      }
+
+      if (moduleName === 'react-native-screens') {
         return {
-          filePath: path.join(
-            __dirname,
-            'node_modules',
-            moduleName,
-            'index.js'
-          ),
+          filePath: path.join(rnsRoot, 'src', 'index.tsx'),
           type: 'sourceFile',
         };
       }
 
-      if (moduleName === 'react-native-safe-area-context') {
-        return {
-          filePath: path.join(
+      if (moduleName in projectPack.dependencies) {
+        for (const ext of resolvedExts) {
+          const possiblePath = path.join(
+            __dirname,
+            'node_modules',
+            moduleName,
+            `index${ext}`
+          );
+
+          const possibleSrcPath = path.join(
             __dirname,
             'node_modules',
             moduleName,
             'src',
-            'index.tsx'
-          ),
-          type: 'sourceFile',
-        };
-      }
+            `index${ext}`
```

Comment:
Why do we want to consider different extensions & paths here? Shouldn't `index.tsx` always be present? What was the issue here?

Reply:
I've mainly coded this for to dynamically resolve any conflict and ensure we're only resolving deps from node_modules, but the thing is that different dependency has different way of storing index file. Some indexes may be included from `src`, and some of them may lie in the main dependency directory. Also, it's not a rule to keep `.tsx` extension for index, since index may have `.js` or `.ts` extension - with this method we're solving most of the cases. I'm aware some of the deps may lie not in `src` dir and not in the main directory aswell, but I don't know specific scenario ¯\_(ツ)_/¯¯\_(ツ)_/¯.

Reply:
Ok, I guess more robust approach would be to compute intersection of app dependencies & react-navigation dependencies & do it only for those. 

As long as it works I don't mind this approach.